token-lending/program/src/processor.rs

```diff
@@ -1176,6 +1184,363 @@ fn process_accrue_interest(program_id: &Pubkey, accounts: &[AccountInfo]) -> Pro
+            return Err(LendingError::InvalidAccountInput.into());
+        }
+    }
+
+    let mut obligation = Obligation::unpack(&obligation_info.data.borrow())?;
+    if obligation_info.owner != program_id {
+        return Err(LendingError::InvalidAccountOwner.into());
+    }
+
+    // @TODO: should this be comparing with deposit_or_withdraw_reserve_collateral_supply_info.key?
+    if &obligation.collateral_reserve != withdraw_reserve_info.key {
+        msg!("Invalid deposit or withdraw reserve account");
+        return Err(LendingError::InvalidAccountInput.into());
+    }
+
+    let obligation_token_mint = unpack_mint(&obligation_token_mint_info.data.borrow())?;
+    if &obligation.token_mint != obligation_token_mint_info.key {
+        msg!("Obligation token mint input doesn't match existing obligation token mint");
+        return Err(LendingError::InvalidTokenMint.into());
+    }
+
+    let obligation_token_input = Token::unpack(&obligation_token_input_info.data.borrow())?;
+    if obligation_token_input_info.owner != token_program_id.key {
+        return Err(LendingError::InvalidTokenOwner.into());
+    }
+    if &obligation_token_input.mint != obligation_token_mint_info.key {
+        return Err(LendingError::InvalidTokenMint.into());
+    }
+
+    // accrue interest and update rates
+    assert_last_update_slot(&borrow_reserve, clock.slot)?;
+    assert_last_update_slot(&withdraw_reserve, clock.slot)?;
+
+    obligation.accrue_interest(borrow_reserve.cumulative_borrow_rate_wads)?;
+
+    let obligation_collateral_amount = obligation.deposited_collateral_tokens;
+    if obligation_collateral_amount == 0 {
+        return Err(LendingError::ObligationEmpty.into());
+    }
+
+    // @TODO: can these next checks be combined? e.g. obligation_collateral_amount <= collateral_amount
+    if obligation_collateral_amount < collateral_amount {
+        // @TODO: does this need a different error type?
+        return Err(LendingError::ObligationEmpty.into());
+    }
+
+    if obligation_collateral_amount == collateral_amount {
+        // @TODO: do obligations ever allow withdrawing all collateral?
+        return Err(LendingError::ObligationEmpty.into());
+    }
+
+    // Check obligation health
+    let trade_simulator = TradeSimulator::new(
+        dex_market_info,
+        dex_market_orders_info,
+        memory,
+        &lending_market.quote_token_mint,
+        // @TODO: are these mint accounts correct?
+        &borrow_reserve.liquidity.mint_pubkey,
+        &withdraw_reserve.liquidity.mint_pubkey,
+    )?;
+
+    // @TODO: this seems simpler than obligation.loan_to_value but is there a tradeoff?
+    let borrow_amount = obligation.borrowed_liquidity_wads.try_ceil_u64()?;
+    // @TODO: should this be using borrow_reserve?
+    let required_collateral_amount = withdraw_reserve.required_collateral_for_borrow(
+        borrow_amount,
+        &borrow_reserve.liquidity.mint_pubkey,
+        trade_simulator,
+    )?;
```

Comment:
This method was intended for borrows only because it uses some special logic to see how much collateral would be needed to execute orders on the dex market to get a certain amount of currency you want to borrow. I think this withdraw logic should probably mirror what we are doing for liquidations where we use the best bid to determine the market price. I think it'd be reasonable to restrict withdraws to keeping the loan about the ltv ratio so that there's still a buffer against the liquidation threshold.

Reply:
Okay, got it. I'll rework this.

Reply:
Okay, I've refactored this in c0b9d7c554b3f4123a5692f9e3eb0664a3fd07c2. Would you check the comments there and tell me if I'm thinking about this correctly?